prov/shm/src/smr_progress.c

```diff
@@ -536,12 +527,13 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 
 	if (cmd->msg.hdr.op == ofi_op_read_req) {
 		hmem_copy_ret = ofi_copy_from_hmem_iov(ptr, cmd->msg.hdr.size,
-						       cmd->msg.data.ipc_info.iface,
-						       device, iov, iov_count, 0);
+					cmd->msg.data.ipc_info.iface,
+					cmd->msg.data.ipc_info.device, iov,
+					iov_count, 0);
 	} else {
 		hmem_copy_ret = ofi_copy_to_hmem_iov(cmd->msg.data.ipc_info.iface,
-						     device, iov, iov_count, 0,
-						     ptr, cmd->msg.hdr.size);
+					cmd->msg.data.ipc_info.device, iov,
```

Comment:
As @shijin-aws correctly pointed out, `cmd->msg.data.ipc_info.device` is somehow messed up - inspected another core dump:
```
(gdb) f 6
#6  0x00007fdc40c873cc in smr_progress_ipc (cmd=0x7fdbfce0a8a0, mr=0x7ffebd0a0330, iov=0x7ffebd0a0350, iov_count=1, total_len=0x7ffebd0a0300, ep=0x2fb39d0, err=0) at prov/shm/src/smr_progress.c:529
529			hmem_copy_ret = ofi_copy_from_hmem_iov(ptr, cmd->msg.hdr.size,
(gdb) p **mr
$7 = {mr_fid = {fid = {fclass = 12, context = 0x0, ops = 0x7fdc40f50800 <ofi_mr_fi_ops>}, mem_desc = 0x3959c70, key = 9437194}, domain = 0x2fb1200, key = 9437194, flags = 70368744177664, iface = FI_HMEM_CUDA, device = 0}
(gdb) p cmd->data
There is no member named data.
(gdb) p cmd->msg
$8 = {hdr = {msg_id = 0, id = 1, op = 2, op_src = 5, op_flags = 6, size = 1, src_data = 262336, data = 0, {tag = 0, {datatype = 0 '\000', atomic_op = 0 '\000'}}}, data = {
    msg = "\001\000\000\000\000\000\000\000\000\000\200\061\311\177\000\000\000\000@\000\000\000\000\000\024\337v\034", '\000' <repeats 12 times>, "\300\267\371\001\000\000\000\000\224K\000\000\000\000\000\000\000\000@\000\000\000\000\000\000\004\000\000\000\000\000\000\035\020\000\000\000\000\000\000\070\000\000\000\000\000\000\000\262\017\320\301\341\000\000\\", '\000' <repeats 95 times>, {iov_count = 1, iov = {{iov_base = 0x7fc931800000, iov_len = 4194304}, {iov_base = 0x1c76df14, iov_len = 0}, {iov_base = 0x1f9b7c0, iov_len = 19348}, {iov_base = 0x400000, iov_len = 1024}, {iov_base = 0x101d, iov_len = 56}, {iov_base = 0x5c0000e1c1d00fb2, iov_len = 0}, {iov_base = 0x0, iov_len = 0},
        {iov_base = 0x0, iov_len = 0}, {iov_base = 0x0, iov_len = 0}, {iov_base = 0x0, iov_len = 0}, {iov_base = 0x0, iov_len = 0}}}, {buf_batch_size = 1, sar = {0, 0, 0, 12672, 32713, 0, 0, 64, 0, 0, -8428, 7286, 0, 0, 0, 0, 0, 0, -18496, 505, 0, 0, 19348, 0, 0, 0, 0, 64, 0, 0, 1024, 0, 0, 0, 4125, 0, 0, 0, 56, 0, 0, 0, 4018, -15920, 225,
        23552, 0 <repeats 18 times>}}, ipc_info = {iface = 1, base_addr = 140502095626240, base_length = 4194304, device = 477552404, offset = 0,
      ipc_handle = "\300\267\371\001\000\000\000\000\224K\000\000\000\000\000\000\000\000@\000\000\000\000\000\000\004\000\000\000\000\000\000\035\020\000\000\000\000\000\000\070\000\000\000\000\000\000\000\262\017\320\301\341\000\000\\\000\000\000\000\000\000\000"}}}
```

The unused `mr` variable contains the correct the device id in this case... The segfault disappears with this change:
```
diff --git a/prov/shm/src/smr_progress.c b/prov/shm/src/smr_progress.c
index 0cda4e53b..344e0fba5 100644
--- a/prov/shm/src/smr_progress.c
+++ b/prov/shm/src/smr_progress.c
@@ -527,9 +527,11 @@ static int smr_progress_ipc(struct smr_cmd *cmd, struct ofi_mr **mr,

        if (cmd->msg.hdr.op == ofi_op_read_req) {
                hmem_copy_ret = ofi_copy_from_hmem_iov(ptr, cmd->msg.hdr.size,
-                                       cmd->msg.data.ipc_info.iface,
-                                       cmd->msg.data.ipc_info.device, iov,
-                                       iov_count, 0);
+                                       (*mr)->iface,
+                                       (*mr)->device,
+                                       iov,
+                                       iov_count,
+                                       0);
        } else {
                hmem_copy_ret = ofi_copy_to_hmem_iov(cmd->msg.data.ipc_info.iface,
                                        cmd->msg.data.ipc_info.device, iov,
```

However I have not got to the root of the location where `cmd->msg.data.ipc_info.device` is set. Any thoughts?

Reply:
... my guess is [here](https://github.com/ofiwg/libfabric/blob/10cdccde6ceeac41eb70cb1dc4eed8f0f8704c02/prov/shm/src/smr_ep.c#L786-L787)

```c

static ssize_t smr_do_ipc(struct smr_ep *ep, struct smr_region *peer_smr, int64_t id,
			  int64_t peer_id, uint32_t op, uint64_t tag, uint64_t data,
			  uint64_t op_flags, enum fi_hmem_iface iface, uint64_t device,
		          const struct iovec *iov, size_t iov_count, size_t total_len,
		          void *context)
{
	struct smr_cmd *cmd;
	struct smr_resp *resp;
	struct smr_tx_entry *pend;
	int ret = -FI_EAGAIN;

	if (ofi_cirque_isfull(smr_resp_queue(ep->region)))
		return -FI_EAGAIN;

	cmd = ofi_cirque_next(smr_cmd_queue(peer_smr));
	resp = ofi_cirque_next(smr_resp_queue(ep->region));
	pend = ofi_freestack_pop(ep->pend_fs);

	smr_generic_format(cmd, peer_id, op, tag, data, op_flags);
	if (iface == FI_HMEM_ZE) {
		if (smr_ze_ipc_enabled(ep->region, peer_smr))
			ret = smr_format_ze_ipc(ep, id, cmd, iov, device,
					total_len, ep->region, resp, pend);
	} else {
		ret = smr_format_ipc(cmd, iov[0].iov_base, total_len, ep->region,
				     resp, iface); // Missing device
	}
...
```

Reply:
Good catch, it seems this PR removes the `device` argument from `smr_progress_*` and derive it from `cmd->msg.data.ipc_info.device`, which is actually not set in the `smr_do_ipc` procedure. Before this change, the `device` was derived from the matched rx_entry's desc so there aws no such issue.

Reply:
I think the quickest fix is to expand the `smr_format_ipc` to have `device` as an argument like `smr_format_ze_ipc` and assign it to `cmd->msg.data.ipc_info.device`. Ideally we should merge `smr_do/format/progress_ipc` with `smr_do/format/progress_ze_ipc` to remove duplication.

Reply:
Thanks for catching. I didn't realize the non-ZE path didn't set the device. We need it in the send IPC path for ZE. I've updated the code to set it in the normal send IPC path so hopefully this should be fixed. The ZE path is unique and needs a special send/recv path but we're working on moving that internally so we don't have to switch by interface in shm. That will let us merge the send and recv code and allow ZE to make use of the IPC caching. Let me know if the update works. Thanks for testing it!

Reply:
Thanks for the quick update. I will test the change today. Once we move past the segfault I'm planning to run AWS CI (i.e. more tests) on a GPU instance. Will report back later.

Reply:
@wenduwan Thanks for the backtrace and additional data. I'm not sure I understand the issue here. I see that the device looks very wrong but I don't understand how it got here or the proposed fix above (I think the formatting might be off).
The problem seems to be on the RMA read path. The copy from is trying to copy the data from the target buffer into the initiator buffer through the IPC handle. The device comes directly from the cmd ipc info which should have been set by the sender. That device comes directly from the user's desc (ofi_mr). In other words, that the device came from the user/efa so I'm not sure how it got such an incorrect device. I think the only two ways this could happen are either with a corruption of the cmd IPC info data or an incorrect desc/MR registration from the app/efa. Can you add some output to see what device that memory region was registered with or what device is being picked up on the send side?
Like on smr_format_ipc what cmd->msg.data.ipc_info.device is at the end of the function and also what device is being passed into that function.
There could be some corruption in the data in the cmd ipc info like when we get the base address and stuff
```
struct ipc_info {
	uint64_t	iface;
	uint64_t	base_addr;
	uint64_t	base_length;
	uint64_t	device;
	uint64_t	offset;
	uint8_t		ipc_handle[MAX_IPC_HANDLE_SIZE];
};
```
Or maybe I'm completely off and your proposed fix is right! I just don't understand what you're suggesting :)

Reply:
@aingerson Thanks for your explanation, in particular this
> The device comes directly from the cmd ipc info which should have been set by the sender. That device comes directly from the user's desc (ofi_mr).

This echoes back to my prior observation around `struct efa_mr` vs `struct ofi_mr`. I have a rough idea to test out via `fi_mr_desc`.

Will sync up later.

Reply:
@aingerson Sorry I just realized that my local PR clone was not up-to-date. I missed git merge failure with your latest commit. I created a new repo and the segfault is gone! It became clear when I read the latest `prov/shm/src/smr_ep.c` 😅
Sorry for the confusion. I will proceed to run additional tests on this PR.